pkg/planner/core/operator/logicalop/logical_union_all.go

```diff
@@ -191,7 +193,46 @@ func (p *LogicalUnionAll) ExhaustPhysicalPlans(prop *property.PhysicalProperty)
 
 // CanPushToCop inherits BaseLogicalPlan.LogicalPlan.<21st> implementation.
 
-// ExtractFD inherits BaseLogicalPlan.LogicalPlan.<22nd> implementation.
+// ExtractFD implement base.LogicalPlan.<22nd> interface.
+func (p *LogicalUnionAll) ExtractFD() *fd.FDSet {
+	// basically extract the children's fdSet.
+	childFDs := make([]*fd.FDSet, 0, len(p.children))
+	for _, child := range p.children {
+		childFD := child.ExtractFD()
+		childFDs = append(childFDs, childFD)
+	}
+	// check the output columns' not-null property.
+	res := &fd.FDSet{}
+	notNullCols := intset.NewFastIntSet()
+	for _, col := range p.Schema().Columns {
+		notNullCols.Insert(int(col.UniqueID))
+	}
+	for _, childFD := range childFDs {
+		notNullCols.IntersectionWith(childFD.NotNullCols)
+	}
+	res.MakeNotNull(notNullCols)
+	// check the equivalency between children.
+	for i := 0; i < p.Schema().Len(); i++ {
+		for j := i + 1; j < p.Schema().Len(); j++ {
+			// detect the equivalency between the i-th and j-th column.
+			flag := true
+			iID := int(p.schema.Columns[i].UniqueID)
+			jID := int(p.schema.Columns[j].UniqueID)
+			for _, childFD := range childFDs {
+				// once we find the i-th and j-th column are not equivalent in one child, we can break the loop.
+				if !childFD.AreColsEquiv(iID, jID) {
+					flag = false
+					break
+				}
+			}
+			if flag {
+				res.AddEquivalence(intset.NewFastIntSet(iID), intset.NewFastIntSet(jID))
+			}
+		}
+	}
```

Comment:
The time complexity is `o(n^2*count(fd from all children))`.
actually we can reduce it to `o(count(fd from all children)` by just iterating the fd from child0 to see calc the intersection with other children